src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Logging/ShowToastNotification.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+/// <summary>
+/// Implements a custom version of the standard 'window/showMessageRequest' to display a toast on the client.
+/// The standard version requires us to wait for a response and then do something on the server with it.
+/// That can be useful in certain cases, but a lot of the time we just want to show a toast with buttons that map to client side commands.
+/// This request allows us to do just that.
```

Comment:
I wonder if this should be part of the standard spec.

Reply:
Possibly, but its been mentioned in other issues that the showMessageRequest was a mistake to add to the spec (because it defines UI).  So not sure if this would get accepted.